Kernel/Graphics/VirtIOGPU/GraphicsAdapter.cpp

```diff
@@ -4,43 +4,53 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/BinaryBufferWriter.h>
 #include <Kernel/Bus/PCI/API.h>
 #include <Kernel/Bus/PCI/IDs.h>
 #include <Kernel/Graphics/Console/GenericFramebufferConsole.h>
 #include <Kernel/Graphics/GraphicsManagement.h>
-#include <Kernel/Graphics/VirtIOGPU/GPU.h>
+#include <Kernel/Graphics/VirtIOGPU/Console.h>
+#include <Kernel/Graphics/VirtIOGPU/FramebufferDevice.h>
 #include <Kernel/Graphics/VirtIOGPU/GraphicsAdapter.h>
 
 namespace Kernel::Graphics::VirtIOGPU {
 
+#define DEVICE_EVENTS_READ 0x0
+#define DEVICE_EVENTS_CLEAR 0x4
+#define DEVICE_NUM_SCANOUTS 0x8
+
 NonnullRefPtr<GraphicsAdapter> GraphicsAdapter::initialize(PCI::DeviceIdentifier const& device_identifier)
 {
     VERIFY(device_identifier.hardware_id().vendor_id == PCI::VendorID::VirtIO);
-    return adopt_ref(*new GraphicsAdapter(device_identifier));
+    auto adapter = adopt_ref(*new GraphicsAdapter(device_identifier));
+    adapter->initialize();
+    return adapter;
 }
 
 GraphicsAdapter::GraphicsAdapter(PCI::DeviceIdentifier const& device_identifier)
-    : PCI::Device(device_identifier.address())
+    : VirtIO::Device(device_identifier)
 {
-    m_gpu_device = adopt_ref(*new GPU(device_identifier)).leak_ref();
-    m_gpu_device->initialize();
+    auto region_or_error = MM.allocate_contiguous_kernel_region(32 * PAGE_SIZE, "VirtGPU Scratch Space", Memory::Region::Access::ReadWrite);
+    if (region_or_error.is_error())
+        TODO();
+    m_scratch_space = region_or_error.release_value();
 }
 
 void GraphicsAdapter::initialize_framebuffer_devices()
 {
-    dbgln_if(VIRTIO_DEBUG, "GPU: Initializing framebuffer devices");
+    dbgln_if(VIRTIO_DEBUG, "GraphicsAdapter: Initializing framebuffer devices");
```

Comment:
It might be a good idea to make these debug prints be something like `VirtIO::GraphicsAdapter` like in the other prints. When looking at the logs, it's nice being able to determine which exact graphics adapter is printing the message.

Reply:
It will be fixed now :)